### PR TITLE
[query] Port ValueToBlockMatrix to BlockMatrixStage2 lowering

### PR DESCRIPTION
**Pull request by chrisvittal**
## Security Assessment
- This change cannot impact the Hail Batch instance as deployed by Broad Institute in GCP

### CODE REVIEW

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
It looks like ValueToBlockMatrix has decent test coverage, but it's mostly indirect (used to test other things) and mostly in python. Could you add a test to BlockMatrixIRSuite that just does some simple sanity checks and edge case checks?

**Comment by chrisvittal** [comment]
I think I got a good mix of Scalar, Array, and NDArray with block sizes that both do and do not divide the shape.

File: hail/hail/src/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala [hunk 1]
```diff
@@ -1145,6 +1145,64 @@ object LowerBlockMatrixIR {
...
+              lowered,
+              MakeTuple.ordered(FastSeq(I64(x.typ.nRows), I64(x.typ.nCols))),
+              True(),
+              ErrorIDs.NO_ERROR,
+            )
+          case _: TNDArray => lowered
+        })
+        val contexts = BMSContexts.tabulate(x.typ, ib) { (rowBlockIdx, colBlockIdx) =>
+          val rowStartIdx = rowBlockIdx.toL * I64(x.typ.blockSize)
+          val colStartIdx = colBlockIdx.toL * I64(x.typ.blockSize)
```
Comment (ehigham) [inline hail/hail/src/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala:1180]: Should these be let-bound to avoid reevaluation?

File: hail/hail/test/src/is/hail/expr/ir/BlockMatrixIRSuite.scala [hunk 2]
```diff
@@ -47,6 +48,45 @@ class BlockMatrixIRSuite extends HailSuite {
     BlockMatrixMap2(left, right, l.name, r.name, ApplyBinaryPrimOp(op, l, r), strategy)
   }
 
+  @Test def testValueToBlockMatrix(): scalatest.Assertion = {
```
Comment (ehigham) [inline hail/hail/test/src/is/hail/expr/ir/BlockMatrixIRSuite.scala]: Please parameterise, make property-based or split these test cases up according to data type as failures in one will hide failures in subsequent assertions.

**Review by ehigham — CHANGES_REQUESTED** [review]
Thanks for adding the tests but I have one request there.
The actual code changes are a nice move/refactor. Without a scala CSE implementation, I wonder if we need to be a little careful about having duplicated computations.

File: hail/hail/test/src/is/hail/expr/ir/BlockMatrixIRSuite.scala (hunk 2 above)
Comment (chrisvittal) [reply hail/hail/test/src/is/hail/expr/ir/BlockMatrixIRSuite.scala]: done

File: hail/hail/src/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala (hunk 1 above)
Comment (chrisvittal) [reply hail/hail/src/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala:1180]: I'm not sure. I did it anyways, it was easy enough.

Comment (ehigham) [reply hail/hail/src/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala:1180]: did you push up that change?

Comment (chrisvittal) [reply hail/hail/src/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala:1180]: yes. https://github.com/hail-is/hail/pull/15086/commits/23ad6e9212bbe73cf6924d7415eb33c5f350571f

Comment (ehigham) [reply hail/hail/src/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala:1180]: Ahhh you used `bindIRs` instead of the ir builder, `ib`! Thanks

**Review by ehigham — APPROVED** [review]

**Review by patrick-schultz — APPROVED** [review]
